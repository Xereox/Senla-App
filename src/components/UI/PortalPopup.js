import React, { useEffect } from 'react'
import { createPortal } from 'react-dom'
import styled from 'styled-components'

const ModalOverFlow = styled.div`
background-color: rgb(0,0,0); /* Fallback color */
background-color: rgba(0,0,0,0.8); /* Black w/ opacity */
position: fixed;
top: 0;
right: 0;
width: 100%;
height: 100%;
overflow: auto;
display: flex;
align-items: flex-start;
justify-content: center;
z-index: 2;
`

const PortalPopup = ({ children }) => {

    let modalRoot = document.getElementById('modal')

    if (!modalRoot) {
        const root = document.getElementById('root')
        modalRoot = document.createElement('div');
        modalRoot.setAttribute('id', 'modal');
        root.after(modalRoot)
    }

    useEffect(function setupElement() {
        return () => (!modalRoot.childElementCount) && modalRoot.remove();
    }, []);

    return createPortal(
        <ModalOverFlow>{children}</ModalOverFlow>,
        modalRoot
    )
}

export default PortalPopup