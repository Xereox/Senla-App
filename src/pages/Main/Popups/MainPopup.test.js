import {screen, render, findAllByRole} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Popup from './MainPopup'
import PortalPopup from "../../../components/UI/PortalPopup";
import Item from "../../../components/UI/CategoryItem";

jest.mock('../../../components/UI/PortalPopup', () => (props) => <div>{props.children}</div>)

const cards = [
    {
        "img": "",
        "title": "Приорбанк",
        "deposit": 315,
        "color": "blue",
        "active": "false",
        "id": "0"
    },
    {
        "img": "",
        "title": "Альфа-банк",
        "deposit": 57,
        "color": "yellow",
        "active": "false",
        "id": "1"
    },
    {
        "img": "",
        "title": "Наличка",
        "deposit": 429,
        "color": "#e09bff",
        "active": "false",
        "id": "2"
    }
]

const props = {
    categoriesCosts: [
        {
            "img": "https://media-exp1.licdn.com/dms/image/C560BAQH9Cnv1weU07g/company-logo_200_200/0/1575479070098?e=2159024400&v=beta&t=QM9VSoWVooxDwCONWh22cw0jBBlBPcBOqAxbZIE18jw",
            "title": "Спорт",
            "color": "#F13C20",
            "id": 1
        },
        {
            "img": "https://media-exp1.licdn.com/dms/image/C560BAQH9Cnv1weU07g/company-logo_200_200/0/1575479070098?e=2159024400&v=beta&t=QM9VSoWVooxDwCONWh22cw0jBBlBPcBOqAxbZIE18jw",
            "title": "Транспорт",
            "color": "#659DBD",
            "id": 2
        },
        {
            "img": "https://media-exp1.licdn.com/dms/image/C560BAQH9Cnv1weU07g/company-logo_200_200/0/1575479070098?e=2159024400&v=beta&t=QM9VSoWVooxDwCONWh22cw0jBBlBPcBOqAxbZIE18jw",
            "title": "Семья",
            "color": "#C38D9E",
            "id": 3
        },
    ],
    currentCategory: 'costs',
    categoriesIncomes: [],
    closePopup: jest.fn(),
    updateContent: jest.fn(),
    setCategory: jest.fn(),
    symbol: 'BR',
    itemInfo: {
        title: 'Some Title',
        img: '',
        IDcategory: 1,
        spending: 1,
        commentary: '',
        IDcard:  5,
        itemDate: ['some date'],
        id: 1
    },
    cards: cards
}

describe('Popup', () => {
    it('without itemInfo button delete should not be in document', () => {
        render(<Popup {...{...props, itemInfo: {...props.itemInfo, id: null}}} />)
        expect(screen.queryByText(/удалить/i)).not.toBeInTheDocument()
    })
    it('with data it have 2 buttons', () => {
        render(<Popup {...props} />)
        expect(screen.queryByText(/удалить/i)).toBeInTheDocument()
        expect(screen.queryByText(/сохранить/i)).toBeInTheDocument()
    })
    it('typing deposit sum should be works', () => {
        render(<Popup {...props} />)
        const depositTextbox = document.querySelector('input:first-of-type')
        userEvent.click(depositTextbox)
        userEvent.clear(depositTextbox)
        userEvent.type(depositTextbox, '422')
        expect(depositTextbox.value).toBe('422')
    })
    it('close function should be works', () => {
        render(<Popup {...props} />)
        userEvent.click(screen.getAllByRole("button")[0])
        expect(props.closePopup).toBeCalled()
    })
    it('category items choose should be works', () => {
        render(<Popup {...props} />)
        const img = document.querySelectorAll(`[data-test-id='img']`)[0]
        userEvent.click(img)
        expect(screen.getAllByText(/спорт/i)[0]).toHaveStyle('color: #ff5200')
    })
    it('should have correct categories items', () => {
        render(<Popup {...props} />)
        const img = document.querySelectorAll(`[data-test-id='img']`)
        expect(img).toHaveLength(props.categoriesCosts.length)
    })
    it('choose cards should be work', () => {
        render(<Popup {...props} />)
        userEvent.click(screen.getByText(/счет/i))
        userEvent.click(screen.getByText(/наличка/i))
        userEvent.click(screen.getByText(/выбрать/i))
        expect(screen.getByText(/наличка/i)).toBeInTheDocument()
    })
    it('change categories onclick should works', () => {
        render(<Popup {...{...props, itemInfo: {...props.itemInfo, id: null}}} />)
        userEvent.click(screen.getByText(/доходы/i))
        expect(props.setCategory).toHaveBeenCalled()
        expect(props.setCategory).toHaveBeenCalledWith('income')
    })
    it('should be like previous snapshot', () => {
        const snap = render(<Popup {...props} />)
        expect(snap).toMatchSnapshot()
    })
})
