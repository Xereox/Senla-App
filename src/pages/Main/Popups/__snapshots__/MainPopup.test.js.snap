// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup should be like previous snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="sc-bkbkJK fnuupP"
        >
          <button
            class="sc-iemWCZ ebNruz"
          >
            X
          </button>
          <div>
            <input
              class="sc-ksluID cSSpoY"
              type="text"
              value="1"
            />
             
            BR
          </div>
          <div
            class="sc-jSFjdj oYTEP"
          >
            <div
              class="sc-gKAaRy edzNoS"
            >
              Счет
            </div>
          </div>
          <label
            class="sc-dIvrsQ ktHHsF"
          >
            Не задан
          </label>
          <div
            class="sc-bqGGPW jKnPDx"
          >
            <div
              class="sc-kEqXSa inkYeI"
            >
              <span
                class="sc-crzoAE iOPuxU"
              >
                Спорт
              </span>
              <div
                class="sc-iqAclL jiVmhW"
                color="#F13C20"
                data-test-id="img"
              />
              <div
                class="sc-dIsUp kVyDxL"
              />
            </div>
            <div
              class="sc-kEqXSa inkYeI"
            >
              <span
                class="sc-crzoAE iOPuxU"
              >
                Транспорт
              </span>
              <div
                class="sc-iqAclL bTSSr"
                color="#659DBD"
                data-test-id="img"
              />
              <div
                class="sc-dIsUp kVyDxL"
              />
            </div>
            <div
              class="sc-kEqXSa inkYeI"
            >
              <span
                class="sc-crzoAE iOPuxU"
              >
                Семья
              </span>
              <div
                class="sc-iqAclL idqmgX"
                color="#C38D9E"
                data-test-id="img"
              />
              <div
                class="sc-dIsUp kVyDxL"
              />
            </div>
          </div>
          <div
            class="sc-fFSPTT bbRVXL"
          >
            <input
              type="date"
              value="2021-07-15"
            />
          </div>
          <div
            class="sc-fnVZcZ hgwRNq"
          >
            Комментарий
            <div>
              <input
                class="sc-hBMUJo jokiav"
                placeholder="Комментарий"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="sc-dlMDgC laVUlp"
          >
            <button
              class="ant-btn ant-btn-text ant-btn-round ant-btn-dangerous"
              type="button"
            >
              <span
                aria-label="delete"
                class="anticon anticon-delete"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
              <span>
                УДАЛИТЬ
              </span>
            </button>
          </div>
          <div
            class="sc-hHEiqL bvvfVj"
          >
            <button
              class="ant-btn ant-btn-primary ant-btn-round ant-btn-block"
              type="button"
            >
              <span>
                Сохранить
              </span>
            </button>
          </div>
          <span
            class="sc-kfYoZR UzCHe"
          >
            <div>
              1. Счёт не выбран
            </div>
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="sc-bkbkJK fnuupP"
      >
        <button
          class="sc-iemWCZ ebNruz"
        >
          X
        </button>
        <div>
          <input
            class="sc-ksluID cSSpoY"
            type="text"
            value="1"
          />
           
          BR
        </div>
        <div
          class="sc-jSFjdj oYTEP"
        >
          <div
            class="sc-gKAaRy edzNoS"
          >
            Счет
          </div>
        </div>
        <label
          class="sc-dIvrsQ ktHHsF"
        >
          Не задан
        </label>
        <div
          class="sc-bqGGPW jKnPDx"
        >
          <div
            class="sc-kEqXSa inkYeI"
          >
            <span
              class="sc-crzoAE iOPuxU"
            >
              Спорт
            </span>
            <div
              class="sc-iqAclL jiVmhW"
              color="#F13C20"
              data-test-id="img"
            />
            <div
              class="sc-dIsUp kVyDxL"
            />
          </div>
          <div
            class="sc-kEqXSa inkYeI"
          >
            <span
              class="sc-crzoAE iOPuxU"
            >
              Транспорт
            </span>
            <div
              class="sc-iqAclL bTSSr"
              color="#659DBD"
              data-test-id="img"
            />
            <div
              class="sc-dIsUp kVyDxL"
            />
          </div>
          <div
            class="sc-kEqXSa inkYeI"
          >
            <span
              class="sc-crzoAE iOPuxU"
            >
              Семья
            </span>
            <div
              class="sc-iqAclL idqmgX"
              color="#C38D9E"
              data-test-id="img"
            />
            <div
              class="sc-dIsUp kVyDxL"
            />
          </div>
        </div>
        <div
          class="sc-fFSPTT bbRVXL"
        >
          <input
            type="date"
            value="2021-07-15"
          />
        </div>
        <div
          class="sc-fnVZcZ hgwRNq"
        >
          Комментарий
          <div>
            <input
              class="sc-hBMUJo jokiav"
              placeholder="Комментарий"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-dlMDgC laVUlp"
        >
          <button
            class="ant-btn ant-btn-text ant-btn-round ant-btn-dangerous"
            type="button"
          >
            <span
              aria-label="delete"
              class="anticon anticon-delete"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="delete"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                />
              </svg>
            </span>
            <span>
              УДАЛИТЬ
            </span>
          </button>
        </div>
        <div
          class="sc-hHEiqL bvvfVj"
        >
          <button
            class="ant-btn ant-btn-primary ant-btn-round ant-btn-block"
            type="button"
          >
            <span>
              Сохранить
            </span>
          </button>
        </div>
        <span
          class="sc-kfYoZR UzCHe"
        >
          <div>
            1. Счёт не выбран
          </div>
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
